### ISSUE TITLE
Stability issues

### ISSUE THREAD

**Issue by maciejhirsz** [not-fl3/good-web-game#15]
Finally [uploaded my stuff to a server](https://game.maciej.codes/). I'm having some weird stability issues:

+ Google Chrome on desktop seems to be handling stuff best and in general has no issues at all.
+ Firefox on desktop is working mostly okay, although lower FPS than Chrome. I did run into an issue with it too though, crashing my entire system with itself, not super fun.
+ iPhone Safari consistently crashes after 2-3 seconds.

Zemeroth works fine for me, so this issue must be due to something I'm doing different, which is either the SpriteBatch or Canvas (Framebuffer). I'll investigate more.

**Comment by not-fl3** [comment]
> I did run into an issue with it too though, crashing my entire system with itself. 

Can confirm :( My system ran out of ram after couple of minutes with the tab opened.

**Comment by not-fl3** [comment]
Still no idea whats going on, but at least I found a way to profile it without crashing my system :) 
With the super narrow window it stop rendering and memory stops leaking for some reason.
![screen](https://user-images.githubusercontent.com/910977/59665198-1d6b6480-91bb-11e9-99ab-819f549c6082.png)

**Comment by not-fl3** [comment]
Just noticed a lot of errors in the console: 
```
Error: WebGL warning: texImage2D: Requested size at this level is unsupported. game.js:643:388
Error: WebGL warning: clear: Framebuffer not complete. (status: 0x8cd6) COLOR_ATTACHMENT0Attachment has no width or height. game.js:703:92
Error: WebGL warning: clear: Framebuffer must be complete. game.js:703:92
Error: WebGL warning: drawArrays: Framebuffer not complete. (status: 0x8cd6) COLOR_ATTACHMENT0Attachment has no width or height
...
Error: WebGL: No further warnings will be reported for this WebGL context. (already reported 32 warnings)
```

**Comment by maciejhirsz** [comment]
@not-fl3 thanks for this! What version of Firefox are you running? I don't see any errors in the console here, I guess iOS safari might suffer from the same issues around FBOs.

**Comment by not-fl3** [comment]
Oh, my bad, the errors was the result of insufficent window size :( 
Also an issue I suppose, but not the one we are looking for.

**Comment by maciejhirsz** [comment]
Ah, that makes sense. Non the less, the memory leak sounds like something that should be possible to track.

**Comment by maciejhirsz** [comment]
Mostly just guessing here, but it looks to me like video ram is leaking, when I look at processes nothing is happening, but the actual memory usage shows a pretty standard GC saw pattern - things go up and down by 2 GB every few seconds in the system monitor:

![](https://maciej.codes/kosz/leak.png)

Zemeroth is showing a completely flat line, so I'm 80% sure that framebuffers (or rather how I'm using them to emulate the Canvas api from ggez) are the culprit here given the rate of bloat here.

**Comment by maciejhirsz** [comment]
DUH!

This is actually related to #14. I have to recreate my main `Canvas` when the screen is resized, but for some reason I figured it would be a good idea to do it as a part of `draw` instead of `resize_event`, and to do that I've also employed a dirty flag to inform the `draw` loop when it needs to create a `Canvas` with a new size. I hate dirty flags but sometimes they are useful. Anyway, everything works fine, except..:

```rust
        if self.canvas_dirty {
            graphics::clear(ctx, [0.0, 0.0, 0.0, 0.0].into());

            self.canvas = Self::create_canvas(ctx, self.canvas_size)?;
        }
```

I never actually set the flag back to `false` :man_facepalming:. Since creating a new `Canvas` creates a new `Texture`, and `Texture` is a leaky abstraction (as per #14) this creates the leak on every frame.
